### PR TITLE
Fix HALIDE_SHARED_LIBRARY=OFF in CMake

### PR DESCRIPTION
**Pull request by steven-johnson**
Building a static library in CMake was totally broken and has been for a long time. This fixes it. Notes:

- Despite the Travis file warning about exceeding the memory limiit, I'm re-enablng it for now for testing purposes

- Uniformly setting -fvisibility=hidden (to match the behavior of Makefile) cascaded into the the annoying DLLEXPORT fixes sprayed across the test directory: because CMake builds these files by compiling to a .o, then linking (rather than compiling+linking the .cpp directly, as Make does); thus we must ensure that the exported functions are marked correctly on posixy systems as well. (Moving this macro elsewhere in a subsequent PR would be fine with me...)

- Minimum CMake bumped to 3.3 to enable the correct behavior of -fvisibility=hidden, because CMake is weird. (That's still a fairly old version.)

- Linking of static libraries in CMake is unspeakably dumb.

### CODE REVIEW

**Comment by abadams** [comment]
The makefile doesn't compile the tests with fvisibility=hidden does it? I thought we restricted that to compiling libHalide itself.

**Comment by steven-johnson** [comment]
Hmm... I thought for sure we did, but it looks like maybe we aren't after all. I can revisit this with that in mind if you like.

**Comment by steven-johnson** [comment]
Pulling out the -fvisibility=hidden from the tests works, but produces horrible linker warnings in a handful of cases:

```
ld: warning: direct access in function 'halide_internal_handle_traits<void, false>::type_info(bool, halide_handle_cplusplus_type::ReferenceType)' from file '../lib/libHalide.a(Pipeline.cpp.o)' to global weak symbol 'guard variable for halide_internal_handle_traits<void, false>::type_info(bool, halide_handle_cplusplus_type::ReferenceType)::the_info' from file 'CMakeFiles/build_correctness_cast_handle.dir/correctness/cast_handle.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```
(repeat a dozen or so times across a few dozen tests)

Investigating.

**Comment by steven-johnson** [comment]
Bah, this is the mixed-visibility problem: in this case, the issue is that the static-globals inside the the templated functions in Type.h end up instantiated inside the test rather than inside libHalide, and the linker doesn't like that.

(Conventional wisdom seems to be "don't mix different -fvisibility settings across multiple libraries of your project; this doesn't bite us in the Makefile version because of luck of linker ordering, I'm guessing.)

**Comment by abadams** [comment]
Doesn't that mean we're causing a headache for everyone using Halide by
using a non-standard default visibility?

**Comment by abadams** [comment]
Basically I'm thinking that the tests should never have to do anything
weird in their source, because user code links to libHalide in exactly the
same way.

**Comment by steven-johnson** [comment]
> Doesn't that mean we're causing a headache for everyone using Halide by
> using a non-standard default visibility?

Well there's the rub: what's the standard? Casual googling makes it look like "use -fvisibility=hidden by default everywhere" is recommended by some people.

We could just back out all this mess by ditching all our EXPORT flags entirely and exporting every symbol (recent versions of MSVC have this as an option).

**Comment by steven-johnson** [comment]
I'm working on a separate PR to try to clean up the static-data-in-templated-functions code in Type.h that should make these warnings moot.

File: Makefile [hunk 1]
```diff
@@ -971,9 +971,13 @@ $(BIN_DIR)/test_internal: $(ROOT_DIR)/test/internal.cpp $(BIN_DIR)/libHalide.$(S
 	$(CXX) $(TEST_CXX_FLAGS) $< -I$(SRC_DIR) $(TEST_LD_FLAGS) -o $@
 
 # Correctness test that link against libHalide
-$(BIN_DIR)/correctness_%: $(ROOT_DIR)/test/correctness/%.cpp $(BIN_DIR)/libHalide.$(SHARED_EXT) $(INCLUDE_DIR)/Halide.h $(RUNTIME_EXPORTED_INCLUDES)
```
Comment (abadams) [inline Makefile]: What is this change about? Is it just to mirror what cmake does, or does the old way actually fail in some way?

Comment (steven-johnson) [reply Makefile]: Good catch -- this was a scalpel left in the patient while I was comparing the two methods. Reverting.

**Comment by steven-johnson** [comment]
Finally clean. PTAL.

**Comment by steven-johnson** [comment]
Finally clean, aside from an unrelated failure in the Windows build. PTAL.

**Comment by abadams** [comment]
What's the status of this, now that we've changed to just exporting everything?

**Comment by steven-johnson** [comment]
It's definitely stale wrt master but could be fixed with modest attention. I'll circle back to it eventually; if anyone else wants to take it and clean it up, that's fine with me.

**Comment by steven-johnson** [comment]
I think the `-fvisibility=hidden` stuff is mostly moot now, so I removed that part (though there may be Windows-related issues I haven't tested yet); assuming this goes green, I think it's ready to review.

(Note that this should ideally be followed up by another PR, to allow us to build both shared and static libraries at the same time, as is done with Makefile; as it is, we'd have to fire off multiple builds to get both, which makes building distribs inconvenient.)

**Comment by steven-johnson** [comment]
OK, notwithstanding https://github.com/halide/Halide/pull/3986, I think this is ready to get a final review and land (pending green), as it does fix the known issues with building static libraries via CMake (at least on Unixy systems; it's possible that there remain some on Windows). 

Most of the fixes here are a subset of https://github.com/halide/Halide/pull/3986, so landing this would actually make that easier, assuming we decide that it's a good idea (which it may or may not be).

**Pull request merged**